### ISSUE TITLE
Allow underscore in `use` bindings

### ISSUE THREAD

**Issue by lostmsu** [fsharp/fslang-suggestions#881]
# Allow underscore in `use` bindings

I propose we allow underscore identifier in `use` bindings to indicate unused value:

```fsharp
use _ = new FileStream("read.me")
//  ^ FS0852: 'use' bindings must be of the form 'use <var> = <expr>'
```

Currently, I have to do this:

```fsharp
use fileLock = new FileStream("read.me") // FS1182: the value is unused
```

or this:

```fsharp
use _fileLock = new FileStream("read.me") // no errors or warnings
```

## Pros and Cons 

Pros: 
- no need to come up with a variable name

Cons: unknown

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S (guess)

Related suggestions: N/A

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by Tarmil** [comment]
@abelbraaksma There's a precedent for allowing `_` but not full patterns though, namely when declaring the "this" variable for a method.

**Comment by lostmsu** [comment]
@jwosty it will be disposed at the end of the block, not necessarily the next line.

**Comment by Happypig375** [comment]
Pretty sure I've seen this declined somewhere before...

**Comment by Happypig375** [comment]
It's another variation of enabling patterns for `use`, marked "probably not": https://github.com/fsharp/fslang-suggestions/issues/693

**Comment by lostmsu** [comment]
@Happypig375 the workaround in #693 does not apply to this scenario.

**Comment by abelbraaksma** [comment]
Instead of writing `_varname`, you can write `__`. It's a legal identifier, maybe that addresses the general issue.

@Tarmil, I know, but in that case it was considered that it wouldn't cause confusion,as it's unlikely programmers would expect the pattern syntax to be valid there.

**Comment by NinoFloris** [comment]
For consistency sake it should just be allowed, it's already allowed in CE context https://github.com/dotnet/fsharp/issues/8570

**Comment by realvictorprm** [comment]
I'm also in favour of this

**Comment by bender2k14** [comment]
In one of my projects, I have `FS1182: the value is unused` configured to be a warning and then any warning configured to cause the build to fail.  If I were to encounter this problem on that project, then I would use this workaround.
```
use __ = ...
__ |> ignore
```

**Comment by cartermp** [comment]
@dsyme thoughts? this one seems very reasonable to me

**Comment by Swoorup** [comment]
Seems useful for applying lock pattern

**Comment by dsyme** [comment]
> @dsyme thoughts? this one seems very reasonable to me


Marked as approved in principle

**Comment by kerams** [comment]
What about `do`-like `use` with no identifier at all? Would it be confusing?

**Comment by abelbraaksma** [comment]
@kerams, `do` requires a unit result. If we follow that logic, it would require an implicit `|> ignore`. I don't think that's a good idea. It would immediately be out of scope and the compiler may consider this an opportunity to immediately dispose the reference.

We don't have a similar construct for `let` either. You cannot presently replace a `let _ = ...` with a `do`, unless you add `|> ignore`, or unless the result is already unit. The latter case cannot happen with disposable objects, as `unit` doesn't implement `IDisposable`, hence `do` or `do!` with disposable objects seems impossible to me.

In short, I think we shouldn't allow an implicit ignore with `do`, it's too different and breaks orthogonality with similar language constructs.

**Comment by Happypig375** [comment]
```fs
open System.IO

let (|Id|) =
    printfn "Hi"
    id
do
    use Id _ = new MemoryStream() // error FS0193: Type constraint mismatch. The type ''a -> MemoryStream' is not compatible with type 'System.IDisposable'
    async {
       use Id _ = new MemoryStream()
       return ()
    } |> Async.Start
```
CEs even allow patterns in `use`, so should we also follow this?

**Comment by Happypig375** [comment]
I wrote an RFC based on this as well as https://github.com/dotnet/fsharp/issues/8570 at https://github.com/fsharp/fslang-design/pull/577.

**Comment by Happypig375** [comment]
RFC discussion here: https://github.com/fsharp/fslang-design/discussions/578

**Comment by dsyme** [comment]
Just to be clear, what's approved is just what's in the title - allowing the use of an underscore. The use of a pattern on the left of `use` for arbitrary code not approved. 

@Happypig375 please file a separate suggestion for allowing pattern matching.  

With regard to pattern matching - I'm not particularly in favour of that, I'll explain why - it's not that it's wrong, it just doesn't solve the actual problem.

People invariable make mistakes when disposing multiple resources.  For example, in the proposed RFC you wrote, sample code like this:
```
    static member Dispose((a:#System.IDisposable, b:#System.IDisposable)) =
        a.Dispose()
        b.Dispose()
```
This is wrong - if the first `Dispose` raises an exception, the second `Dispose` never happens.  It's important to write this correctly.

Other mistakes also happen, notably

    use x, y = Dispose2 (AllocateBigThing1(), AllocateBigThing2())

Here if `AllocateBigThing2()` raises an exception then the resources of `AllocateBigThing1()` are never released except via a finalizer.   This is hard to avoid and really requires a well-written `AllocateTwoBigThings()`  or generic `Disposable2`

    use x, y = Disposable2 (fun () -> AllocateBigThing1()) (fun () -> AllocateBigThing2())

or

    use x, y = Disposable2 AllocateBigThing1 AllocateBigThing2

I think this is the problem that needs solving - making it simpler to allocate multiple resources correctly.  I'm not opposed to solving this problem in language/library, e.g. allowing a tuple on the left, with appropriate correct library helpers.   However I see no real gain in allowing more complex patterns on the left (even if they happen to be allowed in computation expressions).

Also note `using (fun pat -> ...)` can already be used for an arbitary pattern.

**Comment by Happypig375** [comment]
@dsyme What's the return type of `Disposable2`?

**Comment by dsyme** [comment]
> @dsyme What's the return type of Disposable2?

Here, I was imagining a tuple, with your extension member

**Comment by dsyme** [comment]
This was completed here: https://github.com/fsharp/fslang-design/blob/main/FSharp-6.0/FS-1102-discards-on-use-bindings.md
